File: examples/FsharpExamples/Simple.fs
module Simple

// Learn more about F# at http://fsharp.org
open System
open System.Diagnostics
open Pulsar.Client.Api

open Pulsar.Client.Common
open System.Text

let runSimple () =

    let serviceUrl = "pulsar://127.0.0.1:6650"
    let subscriptionName = "my-subscription"
    let topicName = sprintf "my-topic-%i" DateTime.Now.Ticks;


    task {
        let! client =
            PulsarClientBuilder()
                .ServiceUrl(serviceUrl)
                .BuildAsync()

        let! producer =
            client.NewProducer()
                .Topic(topicName)
                .BlockIfQueueFull(true)
                .CreateAsync()

        // let! consumer =
        //     client.NewConsumer()
        //         .Topic(topicName)
        //         .SubscriptionName(subscriptionName)
        //         .SubscriptionType(SubscriptionType.Exclusive)
        //         .SubscribeAsync()

        let bytes = Array.zeroCreate 750
        let sw = Stopwatch()
        sw.Start()
        let n = 10000000
        for i = 0 to n do
            do! producer.SendAndForgetAsync(bytes)
            if i % 100000 = 0 then
                Console.WriteLine($"Sent {i} messages")
        sw.Stop()
        Console.WriteLine($"Sent {n} messages in {sw.ElapsedMilliseconds}ms")

        // let! message = consumer.ReceiveAsync()
        // printfn "Received: %A" (message.Data |> Encoding.UTF8.GetString)
        //
        // do! consumer.AcknowledgeAsync(message.MessageId)
    }

File: examples/CsharpExamples/Simple.cs
using System;
using System.Diagnostics;
using System.Text;
using System.Threading.Tasks;
using Pulsar.Client.Api;

namespace CsharpExamples
{
    internal class Simple
    {
        internal static async Task RunSimple()
        {
            const string serviceUrl = "pulsar://127.0.0.1:6650";
            const string subscriptionName = "my-subscription";
            var topicName = $"my-topic-{DateTime.Now.Ticks}";

            var client = await new PulsarClientBuilder()
                .ServiceUrl(serviceUrl)
                .BuildAsync();
            //
            // var producer = await client.NewProducer()
            //     .Topic(topicName)
            //     .CreateAsync();

            var consumer = await client.NewConsumer()
                .Topic(topicName)
                .SubscriptionName(subscriptionName)
                .SubscribeAsync();

            // var messageId = await producer.SendAsync(Encoding.UTF8.GetBytes($"Sent from C# at '{DateTime.Now}'"));
            // Console.WriteLine($"MessageId is: '{messageId}'");


            var n = 10000000;

            // warmup
            for (var i = 0; i < n / 10; i++)
            {
                var message = await consumer.ReceiveAsync();
                await consumer.AcknowledgeAsync(message.MessageId);
            }

            var sw = new Stopwatch();
            sw.Start();
            for (var i = 0; i < n; i++)
            {
                var message = await consumer.ReceiveAsync();
                await consumer.AcknowledgeAsync(message.MessageId);

                if (i % 100000 == 0)
                    Console.WriteLine($"Received {i} messages");
            }

            sw.Stop();
            Console.WriteLine(
                $"Received {n} messages in {sw.ElapsedMilliseconds}ms. Speed: {n / (sw.ElapsedMilliseconds)}K msg/s");

            // await consumer.AcknowledgeAsync(message.MessageId);
        }
    }
}